Orange/widgets/visualize/owlinearprojection.py

```diff
@@ -1,181 +1,252 @@
-        self._arrow.setPos(self._spine.line().p2())
+class LinearProjectionVizRank(VizRankDialog, OWComponent):
+    captionTitle = "Score Plots"
+    n_attrs = settings.Setting(3)
+    minK = 10
 
-        self._label = TextItem(text=label, color=(10, 10, 10))
-        self._label.setParentItem(self)
-        self._label.setPos(self._spine.line().p2())
+    attrsSelected = Signal([])
+    _AttrRole = next(gui.OrangeUserRole)
 
-    def setLabel(self, label):
-        if label != self._label:
-            self._label = label
-            self._label.setText(label)
+    def __init__(self, master):
+        """
+        Add the spin box for a number of attributes to take into account.
+        """
+        VizRankDialog.__init__(self, master)
+        OWComponent.__init__(self, master)
 
-    def setPen(self, pen):
-        self._spine.setPen(pen)
+        box = gui.hBox(self)
+        self.n_attrs_spin = gui.spin(
+            box, self, "n_attrs", 2, 4,
+            label="Number of attributes to take into account: ",
+            controlWidth=50, alignment=Qt.AlignRight,
+            callback=self._n_attrs_changed)
+        gui.rubber(box)
 
-    def paint(self, painter, option, widget):
-        pass
+        self.last_run_n_attrs = None
 
-    def boundingRect(self):
-        return QRectF()
+    def before_running(self):
+        """
+        Disable the spin for number of attributes before running and
+        enable afterwards. Also, if the number of attributes is different than
+        in the last run, reset the saved state (if it was paused).
+        """
+        if self.n_attrs != self.last_run_n_attrs:
+            self.saved_state = None
+            self.saved_progress = 0
+        if self.saved_state is None:
+            self.scores = []
+            self.rank_model.clear()
+        self.last_run_n_attrs = self.n_attrs
 
-    def viewTransformChanged(self):
-        self.__updateLabelPos()
+    def check_preconditions(self):
+        if not super().check_preconditions():
+            return False
+        if not self.master.data:
+            return False
+        if not self.master.data.domain.class_var:
+            return False
+        self.n_attrs_spin.setMaximum(self.master.n_cont_var)
+        return True
+
+    def state_count(self):
+        n_all_attrs = len(self.attrs)
+        n_attrs = self.n_attrs
+        return ceil(0.5 * factorial(n_all_attrs) / (factorial(n_all_attrs - n_attrs) * n_attrs))
+
+    def iterate_states(self, state):
+        if state is None:  # on the first call, compute order
+            attrs = self._score_heuristic()
+            self.attrs = attrs
+        for c in combinations(attrs, self.n_attrs):
+            for p in islice(permutations(c[1:]), ceil(factorial(len(c) - 1)/2)):
+                yield (c[0], ) + p
+
+    def compute_score(self, state):
+        master = self.master
+        variables = [s for s in state]
+        master.prepare_plot_data(variables)
```

Comment:
This call essentially resets the widget!? Take brown-selected data, start computation and, while it is computing, click on one of the suggested projection. The widget will of course crash since VizRank messes it up.

This is bad design.